File: razer-sdk/src/main/java/net/booone/razersdk/sdk/RzKeyArray.java
package net.booone.razersdk.sdk;

import java.util.Arrays;
import java.util.Optional;

public final class RzKeyArray {
    private static final int MAX_ROW = 5;
    private static final int MAX_COLUMN = 21;

    private static final RzKey[][] RZ_KEY_ARRAY = new RzKey[MAX_ROW + 1][MAX_COLUMN + 1];

    static {
        Arrays.stream(RzKey.values()).forEach(rk -> RZ_KEY_ARRAY[rk.getRow()][rk.getColumn()] = rk);
    }

    public static Optional<RzKey> keyAt(int row, int column) {
        if (row > MAX_ROW || row < 0 || column > MAX_COLUMN || column < 0) {
            return Optional.empty();
        }
        return Optional.ofNullable(RZ_KEY_ARRAY[row][column]);
    }
}

File: src/razer-sdk/sdk/RzKeyArray.ts
import {getColumn, getRow, getRzKeyValues, RzKey} from './RzKey.js';

const MAX_ROW = 5;
const MAX_COLUMN = 21;

const RZ_KEY_ARRAY: (RzKey | undefined)[][] = Array.from(
    { length: MAX_ROW + 1 },
    () => new Array<RzKey | undefined>(MAX_COLUMN + 1).fill(undefined),
);

// Static initialization - populate the lookup array
for (const rk of getRzKeyValues()) {
    RZ_KEY_ARRAY[getRow(rk)][getColumn(rk)] = rk;
}

export function keyAt(row: number, column: number): RzKey | undefined {
    if (row > MAX_ROW || row < 0 || column > MAX_COLUMN || column < 0) {
        return undefined;
    }
    return RZ_KEY_ARRAY[row][column];
}
